network/transport/v2/gossip/queue.go

```diff
@@ -0,0 +1,162 @@
+/*
+ * Nuts node
+ * Copyright (C) 2022 Nuts community
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+package gossip
+
+import (
+	"container/list"
+	"context"
+	"sync"
+	"time"
+
+	"github.com/nuts-foundation/nuts-node/crypto/hash"
+)
+
+// peerQueue contains a log of received transaction references and a queue of references to send for a specific peer.
+type peerQueue struct {
+	// cancelFunc to stop the ticker for this peer
+	cancelFunc context.CancelFunc
+	// log of recent received transaction hashes
+	log *list.List
+	// logSet contains the same set as log, but then as indexed set.
+	logSet map[string]*list.Element
+	// maxSize determines how many items are kept in each list
+	maxSize int
+	// mutex to protect concurrent access to the queue
+	mutex sync.Mutex
+	// queue that holds the list of hashes to be gossipped
+	queue *list.List
+	// set contains the items from the queue but as indexed set
+	set map[string]*list.Element
+}
+
+func newPeerQueue() peerQueue {
+	return peerQueue{
+		log:     list.New(),
+		logSet:  map[string]*list.Element{},
+		maxSize: maxQueueSize,
+		queue:   list.New(),
+		set:     map[string]*list.Element{},
+	}
+}
+
+// start a ticker. It'll use the given context as parent context to stop the ticker
+func (pq *peerQueue) start(parentCtx context.Context, interval time.Duration) <-chan bool {
+	var ctx context.Context
+	ctx, pq.cancelFunc = context.WithCancel(parentCtx)
+	done := ctx.Done()
+	returnChan := make(chan bool, 1)
+
+	go func() {
+	outer:
+		for {
+			select {
+			case <-done:
+				break outer
+			case <-time.Tick(interval):
```

Comment:
Note that we should probably use `time.Ticker` here instead (as `time.Tick` leaks the underlying ticker)